Sources/BraintreeCoreSwift/BTClientToken.swift

```diff
@@ -0,0 +1,118 @@
+
+    // MARK: - Public Properties
+
+    /// The client token as a BTJSON object
+    public let json: BTJSON
+
+    /// The extracted authorization fingerprint
+    public let authorizationFingerprint: String
+
+    /// The extracted configURL
+    public let configURL: URL
+
+    /// The original string used to initialize this instance
+    public let originalValue: String
+
+    // MARK: - Initializers
+
+    /// Initialize a client token with a client token string generated by a Braintree Server Library.
+    /// - Parameter clientToken: A client token string generated by a Braintree Server Library
+    @objc(initWithClientToken:error:)
+    public init(clientToken: String) throws {
+        // Client token must be decoded first because the other values are retrieved from it
+        self.json = try Self.decodeClientToken(clientToken)
+        
+        guard let authorizationFingerprint = json["authorizationFingerprint"].asString(),
+              !authorizationFingerprint.isEmpty else {
+            throw BTClientTokenError.invalidAuthorizationFingerprint
+        }
+        
+        guard let configURL = json["configUrl"].asURL() else {
+            throw BTClientTokenError.invalidConfigURL
+        }
+        
+        self.authorizationFingerprint = authorizationFingerprint
+        self.configURL = configURL
+        self.originalValue = clientToken
+    }
+    
+    // MARK: - Internal helper functions
+
+    private static func decodeClientToken(_ rawClientToken: String) throws -> BTJSON {
+        let data: Data
+        let isBase64: Bool
+        if let base64Data = Data(base64Encoded: rawClientToken) {
+            data = base64Data
+            isBase64 = true
+        } else if let utf8Data = rawClientToken.data(using: .utf8) {
+            data = utf8Data
+            isBase64 = false
+        } else {
+            throw BTClientTokenError.failedDecoding("Base64 or UTF8 encoding is required for Client Token.")
+        }
+
+        guard let clientTokenJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
+            throw BTClientTokenError.invalidFormat("Invalid JSON. Expected to find an object at JSON root.")
+        }
+
+        guard let version = clientTokenJSON["version"] as? Int else {
+            throw BTClientTokenError.invalidFormat("Invalid version number. Expected to find an integer for key \"version\".")
+        }
+
+        // Version 1 must be utf8, versions 2 & 3 must be base64
+        switch version {
+        case 1:
+            if isBase64 {
+                throw BTClientTokenError.failedDecoding("UTF8 encoding is required for Client Token version 1.")
+            }
+
+        case 2, 3:
+            if !isBase64 {
+                throw BTClientTokenError.failedDecoding("Base64 encoding is required for Client Token versions 2 & 3.")
+            }
+            
+        default:
+            throw BTClientTokenError.unsupportedVersion
+        }
```

Comment:
This is a take-it-or-leave-it comment, we could avoid a level of nesting without using a switch. Whichever you find more readable.

```suggestion
        if version == 1 && isBase64 {
            throw BTClientTokenError.failedDecoding("UTF8 encoding is required for Client Token version 1.")
        } else if (version == 2 || version == 3) && !isBase64 {
            throw BTClientTokenError.failedDecoding("Base64 encoding is required for Client Token versions 2 & 3.")
        } else if (version > 3) {
            throw BTClientTokenError.unsupportedVersion
        }
```

Reply:
Yeah, I think this is a lot cleaner than a nested if in a switch. I think in the final else if we probably want to do `else if version < 1 || version > 3` to catch any unlikely weirdness in the `Int` we get back from the guard above this.